GA4 Importer

AltamashShaikh opened the pull request:
### Description:

Importer to import GA4 Data into Matomo.
Fixes: #150 
works with existing OAuth credentials
GA Data API and GA Admin API access is required
The GA Data and Admin API are still in beta phase

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

AltamashShaikh committed:
Initial code push

AltamashShaikh committed:
Added code to import VisitsSummary

AltamashShaikh committed:
Code to calculate bounce_count metric

AltamashShaikh committed:
support for bounces_count added via bounce_Rate

AltamashShaikh commented:
Unsupported/Incompatible Doc
```
ga:exits, ga:entrances, ga:pageDownloadTime, ga:pageLoadSample, ga:goal{$gaIdGoal}Completions, ga:goal{$gaIdGoal}Value, ga:goal{$gaIdGoal}ConversionRate

ga:bounces(Bounces) is no more supported in GA4 but can be calculated using the bounceRate metric (bounceRate * sessions)

No Conditions for goals to fire, as goals are converted to events.
Couldn't find anything for segments in GA4 yet.

AltamashShaikh commented on ImporterGA4.php:
```diff
@@ -0,0 +1,698 @@
+            if (!method_exists(SettingsServer::class, 'isMatomoForWordPress') || !SettingsServer::isMatomoForWordPress()) {
+                $idSite = SitesManagerAPI::getInstance()->addSite(
+                    $siteName = $webProperty->getDisplayName(),
+                    $urls = $type === \Piwik\Plugins\MobileAppMeasurable\Type::ID ? null : [$webProperty->getDisplayName()],
```
API doesn't support any method to fetch the URL

AltamashShaikh commented on ImporterGA4.php:
```diff
@@ -0,0 +1,698 @@
+                $idSite = SitesManagerAPI::getInstance()->addSite(
+                    $siteName = $webProperty->getDisplayName(),
+                    $urls = $type === \Piwik\Plugins\MobileAppMeasurable\Type::ID ? null : [$webProperty->getDisplayName()],
+                    $ecommerce = 1,
```
No additional setting to fetch if its Ecommerce since its event based now, it supports ecommerce directly.

AltamashShaikh committed:
Added RecordImporter for MarketingCampaignsReporting

AltamashShaikh committed:
Merged with 4.x-dev

AltamashShaikh committed:
Added Importer for Referrers

AltamashShaikh committed:
Added Importer for Actions

AltamashShaikh committed:
Added Importer for DevicesDetection

AltamashShaikh committed:
Added Importer for Customdimension

AltamashShaikh committed:
Changes to support Oauth instead of service account

AltamashShaikh committed:
Added runreport as a batch as max 10 metrics are supported at a time and added Events Importer

AltamashShaikh committed:
Added Goals,Resolution,UserCountry and UserLanguage Importers

AltamashShaikh committed:
Started using languageCode dimension directly

AltamashShaikh committed:
Added importer for VisitTime

AltamashShaikh committed:
Created Ui to schedule GA4 import

AltamashShaikh committed:
Fixed integration test case

AltamashShaikh committed:
Merged with 4.x-dev

AltamashShaikh committed:
Chnaged reordering by id desc and fixed table width issue

AltamashShaikh committed:
Fixed customDimensions for GA4

AltamashShaikh committed:
Fixed ImportStatusTest

AltamashShaikh committed:
Added ImportStatusTest cases for GA4

AltamashShaikh committed:
Hanadled notices

AltamashShaikh committed:
Updated UI screenshots

AltamashShaikh committed:
Added CustomVariables and VisitorInterest Importer

AltamashShaikh committed:
Started importing GA4 goals as event goals

AltamashShaikh committed:
GA selector UI as per moc created

AltamashShaikh committed:
Updated test cases

AltamashShaikh committed:
Updated UI for GA import as per new mock and fixed and added new testcases for GA4

AltamashShaikh committed:
Fixed UI test cases

AltamashShaikh committed:
Added system test case for GA4 and completed testcapture for GA4

AltamashShaikh committed:
Formatting changes and added ini_set('memory_limit', '-1'); in MockRestTransport.php

AltamashShaikh committed:
Updated screenshots

Chardonneaur commented:
@AltamashShaikh  does it include GA row data? Could we get the log visit report thanks to it?

AltamashShaikh commented:
@Chardonneaur I didn't understand the request sorry, can you give more details about it ?

AltamashShaikh committed:
Changed for UI test to work for GA4

Chardonneaur commented:
@AltamashShaikh people claim that GA 4 allow you to access to everything, does it mean that we will get the log visit report filled with data with GA 4 importer?

AltamashShaikh commented:
@Chardonneaur I will check if we can add it, but mostly that won't be possible as we migrate the aggregated data and not the granular data for GA Import.

AltamashShaikh committed:
Created Common MockResponseBuilderGA4 to save memory for responses and UI test added timeout to test

AltamashShaikh committed:
Removed ini_set from MockTransport and waitForTimeout from GA4 ui test

AltamashShaikh committed:
changes for debug

AltamashShaikh committed:
added more option for mock debug

AltamashShaikh committed:
Mock response not found fix

AltamashShaikh committed:
Removed log functions and updated tests

AltamashShaikh committed:
UI test fixed and captured-ga4-log file updated without x-goog-api-client header

AltamashShaikh committed:
Added unit test cases for GA4

Chardonneaur commented:
@AltamashShaikh is GA4 blocking the export of granular data?

AltamashShaikh commented:
@Chardonneaur There are row limits, [Refer](https://support.google.com/analytics/answer/10702008?hl=en)

AltamashShaikh committed:
Added unit test for GoogleGA4ResponseDataTableFactory

AltamashShaikh committed:
Changes for Visti frequency record importer added

AltamashShaikh committed:
Fixed test cases

AltamashShaikh committed:
Merge branch '3.x-dev' into 150-GA4-Import

AltamashShaikh committed:
Merge branch '4.x-dev' into 150-GA4-Import

AltamashShaikh committed:
Merged with 4.x-dev

AltamashShaikh committed:
Added code to import from most recent dates first for GA4

AltamashShaikh committed:
Updated UI screenshot

AltamashShaikh committed:
Added Intergration test for GA4 Importer

MarcoPNS commented:
Hey! Just wanted to test the current state of the GA4 Import. Is it correct that the import runs backwards?

<img width="909" alt="Bildschirmfoto 2022-08-16 um 12 42 33" src="https://user-images.githubusercontent.com/35105681/184860864-4bd2190b-1fc8-4264-a09b-ffc9feb8dd52.png">

AltamashShaikh commented:
@MarcoPNS  Yes the importer will try to import the most recent dates first, earlier it used to start from `start_date` and end with `end_date` now we try to import the recent dates first and today and other future dates at last

justinvelluppillai commented:
@MarcoPNS this helps imports with a long date range as it can otherwise take days or weeks for the data to be available, so it is useful to have more recent data first. If this is a problem for your use case it'd be good to hear more about this to inform our future efforts.

MarcoPNS commented:
I don't have a problem with it personally, but I don't think it's implemented properly yet, is it? Because currently, this does not work sooo great for me, unfortunately. The status does not update and new imports do not survive a rate limit. They change the status to "finished" because status and end-date is the same. 

So I changed the line to the previous for-loop so that I can import everything again:

https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/d23b1b80355b887019ed88fdd6cbf981889d8e3f/ImporterGA4.php#L397-L398

AltamashShaikh commented:
Hey @MarcoPNS can you give more detail on the issue you faced?
It will help us to replicate and fix the issue

MarcoPNS commented:
Sure! I use Matomo 4.11.0 - I used the current live version of the GA-Importer. Then I removed it, cleared the cache and installed the 150-GA4-Import Branch manually. Then I tried to import a GA4 Property with following dates: 2021-08-16, 2022-08-16

Last imported date stays 2022-08-16 all the time. Last archieved date stays empty. After reaching the rate-limit and restarting the status changes to killed or finished.

Do such if-queries perhaps have something to do with it?
https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/f261eff944593b869f620253171c5e3141d7a516/Importer.php#L437-L439

https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/77c3f61b7e09890a7df6fe583fbba379caeb2184/Tasks.php#L182-L187

AltamashShaikh commented:
@MarcoPNS I tried replicating but works fine for me may be the property I am testing doesn't have much data, can you please do this below steps to help us in debugging the issue ?

Revert the changes you did in [plugin-GoogleAnalyticsImporter/ImporterGA4.php](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/d23b1b80355b887019ed88fdd6cbf981889d8e3f/ImporterGA4.php#L397-L398)

Configure a new import with verbose logging enabled
![Screenshot from 2022-08-18 13-45-44](https://user-images.githubusercontent.com/14902277/185345645-2245e4c2-599a-4e22-9ac4-b3b42f41e1f3.png)

There will be a log file created under /path/to/matomo/tmp/logs/ with a filename like `gaimportlog.{idSite_new_configured_account}..log`

Can you share that file ?

MarcoPNS commented:
I left the import a little out of sight. When I looked again, the import was already "finished". 

<img width="1267" alt="Bildschirmfoto 2022-08-18 um 10 50 04" src="https://user-images.githubusercontent.com/35105681/185355446-3d0fd72d-2a46-4228-a6ac-9ce3e3177dc7.png">

However, the import has again survived only till the rate limit.

<img width="918" alt="Bildschirmfoto 2022-08-18 um 10 50 49" src="https://user-images.githubusercontent.com/35105681/185354178-a385fab5-aa4f-47a2-a432-2648d174f52b.png">

[gaimportlog.log](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/files/9372021/gaimportlog.log)

AltamashShaikh commented:
@MarcoPNS Thank you very much.
I can see we have an unhandled case for hourly limit.
Will add a fix and hopefully this should work

AltamashShaikh committed:
Updated code to handle Hourly and Daily limit reached

AltamashShaikh commented:
@MarcoPNS Thanks for the log file, we just updated the code to handle the hourly rate limit exception.
This should not killed the import and eventually take alot more time since it will back of for an hour.

You need to update the plugin with newest change
Configure a new import and please configure a new import with verbose logging enabled so that we can check if its still breaking.

MarcoPNS commented:
Hi! Its still breaking: [gaimportlog.log](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/files/9393141/gaimportlog.log)

AltamashShaikh committed:
Updated Exception throw condition for daily andhourly limit

AltamashShaikh commented:
@MarcoPNS I had added additional logging to know the status, but that also didn't worked.
I just updated the condition to catch the exception by `status` too.
Can we repeat the activity of updating the code and running a new import one more time ?

MarcoPNS commented:
Here's the new log file:
[gaimportlog.log](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/files/9394705/gaimportlog.log)

AltamashShaikh commented:
@MarcoPNS With the recent code changes, I can see its getting "rate_limited"
![Screenshot from 2022-08-23 11-20-47](https://user-images.githubusercontent.com/14902277/186080911-fc42e553-00e9-4d83-ba79-215f4de832f3.jpg)

Are you sure you have update the correct code and branch ?

AltamashShaikh committed:
Added code for rate_limited_hourly import status

AltamashShaikh commented:
@MarcoPNS Also updated the code to display "rate_limited_hourly" as a status instead of "rate_limited" for hourly rate limitation exception.
![Screenshot from 2022-08-23 11-20-47-hourly](https://user-images.githubusercontent.com/14902277/186082966-cc51421f-f972-4d12-a163-700fca5c7d5e.jpg)

AltamashShaikh commented:
@MarcoPNS After an hour the import ran and was finished successfully
![Screenshot from 2022-08-23 12-32-03](https://user-images.githubusercontent.com/14902277/186092704-4767c207-ae5a-461f-84ac-6c6c30f91c91.png)

MarcoPNS commented:
Hi! I always disabled the plugin beforehand, cleared the cache, deleted the folder and then pulled and uploaded the latest version of your branch. Unfortunately, the status still jumps directly to finished although a lot of data is still missing.

If you need a good filled GA4 property to test this better maybe a [GA4 demo account](https://support.google.com/analytics/answer/6367342) will help.

AltamashShaikh committed:
Exception code updated

AltamashShaikh commented:
@MarcoPNS The demo account has no access to API unfortunately.
Getting this error.
![Screenshot from 2022-08-24 13-48-13](https://user-images.githubusercontent.com/14902277/186367611-994c6918-5d34-44aa-8423-8f78d61221d0.png)

AltamashShaikh commented:
I tried hitting the limit for hour and I could see it working properly after the next hour.
I recently made few changes, you can update and check if its working. Also can you share what status do you see in the UI ?
Is it `killed`, `rate_limited` or `finished` ?
